### ISSUE TITLE
nonzero function in C++ much slower than python(CUDA) 

### ISSUE THREAD

**Issue by otakuxiang** [pytorch/pytorch#39649]
I find that nonzero function in C++ API much slower than python.
Here comes my test code:

```
auto test = torch::ones({1,368,1224});
    for(int i = 0;i < 368;++i){
        for(int j = 0;j<1224;++j){
            float f;
            fin>>f;
            test.index_put_({0,i,j},f);
        }
    }    //just read the a tensor from file 
    test = test.cuda();
    clock_t total = 0;
    for(int i = 0;i<100;++i){
        clock_t startTime = clock();
        auto t = at::nonzero(test);
        clock_t endTime = clock();
        total += endTime-startTime;
    }

    cout<<(double)(total)/CLOCKS_PER_SEC * 1000 / 100<<endl;
```

and python code:
```
test = torch.Tensor(test).cuda()
total = 0.0
for i in range(100):
    torch.cuda.synchronize()
    start = time.time()
    t = torch.nonzero(test)
    torch.cuda.synchronize()
    end = time.time()
    total += end-start

print((end-start)*10,"ms")
```

C++ result is: 0.32987 ms
python result is : 0.0018024444580078125 ms

The env of my computer:
```
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1660 Ti
Nvidia driver version: Could not collect
cuDNN version: Could not collect
```
the libtorch bulid-version is 1.6.0.dev20200603+cu101

cc @yf225 @glaringlee @VitalyFedyunin @ngimel

**Comment by glaringlee** [comment]
@otakuxiang 
I did a run on my machine. It was 0.501 ms in c++ vs 0.519 ms in python.
I use pytorch 1.4 for python and libtorch 1.5 for c++.
c++ code is like this, python is the same:
```
long long total = 0;
for(int i = 0;i<100;++i){
    auto start = std::chrono::high_resolution_clock::now();
    auto t = at::nonzero(test);
    auto elapsed = std::chrono::high_resolution_clock::now() - start;
    total += std::chrono::duration_cast<std::chrono::microseconds>(elapsed).count();
}
cout << (double)total/1000/100 << endl;
```


Something to point out here:
1. clock() is a very low precision, if you want to test running time in less than 10ms, please use some other apis with higher precision. for eg. gettimeofday under linux, or std::high_resolution_clock in c++ libs.

2. The running time calculation in your python code is print((end-start)*10,"ms"), correct me if I am wrong, it should be *1000 since time.time() return seconds.

**Comment by otakuxiang** [comment]
@glaringlee 
Thank you for your answer! I just use the code you provide and the result is just like which using clock() 0.32847 ms.
For python code, I know end - start return seconds. I just chage 
`
print((end-start)*1000 /100,"ms")   #nonzero runs 100 times
`
into 
`
print((end-start)*10,"ms")   #nonzero runs 100 times
`
So, actually in your computer, nonzero in C++ also 100x slower than python.

**Comment by glaringlee** [comment]
@otakuxiang 
No, end and start is for one loop running time, you shouldn't divide by 100.
If you want to divide by 100, you should use total.

**Comment by otakuxiang** [comment]
@glaringlee 
Thanks! But when it I use the nonzero function to post process the tensor, it performs worse.Here comes my real code
`

  int batch = images.size();
  int H = images[0].Height(),W = images[0].Width();
  std::vector<torch::Tensor> inputt(batch);
  #pragma omp parallel for
  for(int i = 0;i < batch;++i){
    cv::Mat img;
    cv::eigen2cv(images[i].GetMat(), img);
    img.convertTo(img,CV_32F,1/255.0);
    inputt[i] = torch::from_blob(img.data,{1,1,H,W}).cuda();
  }
  
  torch::Tensor input_tensor = torch::cat(inputt,0);
  std::vector<torch::jit::IValue> inputs;
  inputs.push_back(input_tensor);
  torch::NoGradGuard no_grad;
  auto output = module.forward(inputs).toTuple();
  auto score = output->elements()[0].toTensor();
  auto descriptor = output->elements()[1].toTensor();
  #pragma omp parallel for
  for(int i = 0;i<batch;++i){
    torch::NoGradGuard no_grad;
    auto startTime = std::chrono::high_resolution_clock::now();
    auto kp = at::nonzero(score.index({i}) > 0.005);
    auto endTime = std::chrono::high_resolution_clock::now();
    cout<<(double)(std::chrono::duration_cast<std::chrono::microseconds>(endTime-startTime).count())/1000<<endl;   // this print 25.475 ms
    auto mask = ((kp.index({Slice(None,None),0}) >= 4) & (kp.index({Slice(None,None),0}) < (H - 4))) 
      & ((kp.index({Slice(None,None),1}) >= 4) & (kp.index({Slice(None,None),1}) < (W - 4)));
    
    kp = kp.index({mask,Slice(None,None)});
    kp = at::flip(kp,{1}).toType(torch::kFloat32);
    
    auto div = torch::zeros({1,1,1,2});
    div.index_put_({0,0,0,0},W-4.5);
    div.index_put_({0,0,0,1},H-4.5);
    auto kps = (kp-3.5).reshape({1,1,-1,2})/div.cuda();
    kps = kps * 2 - 1;
    
    auto desc = torch::nn::functional::grid_sample(
        descriptor.index({Slice(i,i+1),Slice(None,None),Slice(None,None),Slice(None,None)}),kps, 
        torch::nn::functional::GridSampleFuncOptions().mode(torch::kBilinear).padding_mode(torch::kZeros).align_corners(true));
    desc = torch::nn::functional::normalize(desc.reshape({256,-1}),
      torch::nn::functional::NormalizeFuncOptions().p(2).dim(0));
    keyPoints[i] = kp.cpu().contiguous();
    descs[i] = desc.cpu().contiguous();
  }
`
However in python:
`
def forward(self, scores , descriptors):
       torch.cuda.synchronize()
        start = time.time()
        keypoints = torch.nonzero(scores > self.keypoint_threshold)
        torch.cuda.synchronize()
        end = time.time()
        print("nonzero use:",(end-start)*1000) 
        
        h, w = scores.shape
        mask_h = (keypoints[:, 0] >= self.remove_borders) & (keypoints[:, 0] < (h - self.remove_borders))
        mask_w = (keypoints[:, 1] >= self.remove_borders) & (keypoints[:, 1] < (w - self.remove_borders))
        mask = mask_h & mask_w
        keypoints = keypoints[mask,:]
        keypoints = torch.flip(keypoints, [1]).float()
       
        c, h, w = descriptors.shape
        kp = keypoints - 3.5
        kp /= torch.tensor([(w*8 - 8/2 - 0.5), (h*8 - 8/2 - 0.5)],
                              ).to(keypoints)[None]
        kp = kp*2 - 1  # normalize to (-1, 1)   
        descriptors = torch.nn.functional.grid_sample(
        descriptors.reshape(1,c,h,w), kp.reshape(1, 1, -1, 2), mode='bilinear', align_corners=True)
        descriptors = torch.nn.functional.normalize(
            descriptors.reshape(1,c, -1), p=2., dim=1)
        
        return descriptors[0],keypoints
`
It prints nonzero use: 0.2684593200683594

**Comment by otakuxiang** [comment]
Something more interesting is that the sleep time add nonzero time always equals 27ms.
I guess the problem is in C++ I don't call some function like torch.cuda.synchronize(), so is there any function in torch C++ API does the same as torch.cuda.synchronize()?

**Comment by glaringlee** [comment]
@otakuxiang 
I don't think sleep matters. If you can post a runnable code, I can run it on my machine and check. I noticed that the c++ code and python code use different tensor indexing when calling at::nonzero. It shouldn't cause that much difference though. Since I can not see your full runable code, pls make sure that the tensors used in two codes have exactly the same shape and device type. 

Also, can you replace the score.index({i}) > xxxx with a pre-defined tensor, so it will show whether the tensor indexing caused the slowness.

**Comment by otakuxiang** [comment]
@glaringlee
Thanks.  
It's right that sleep doesn't matters.But it really helps me to find the problem. I find the problem is in C++ model infer is also asynchronous, so it's wrong to measure nonzero function without synchronize with cuda after the model inferrence. I will write a new issue about how to call cudaDeviceSynchronize() in torch C++.

**Comment by xuezu29** [comment]
> @glaringlee
> Thanks.
> It's right that sleep doesn't matters.But it really helps me to find the problem. I find the problem is in C++ model infer is also asynchronous, so it's wrong to measure nonzero function without synchronize with cuda after the model inferrence. I will write a new issue about how to call cudaDeviceSynchronize() in torch C++.

I met the same problem. Have you solved it?

**Comment by otakuxiang** [comment]
First, when you measure the model running time in C++ code,you need call cudaDeviceSynchronize() function before you call time function such as clock().
Second, if you use .pt model, there is nothing wrong that program run slower than python when you first call foward() function.Just call foward() function again,you will find that speed as fast as it in python.
If you don't mind it,you can call foward() function in initialize function after init the model.

**Comment by xuezu29** [comment]
@otakuxiang 
Thanks . I know the first  call foward() will take longer，I load the python trained model by torch::jit::load() ，C++ takes 60-70 ms longer than python. Is this normal?

**Comment by otakuxiang** [comment]
No, it's not normal, can you show your time measuring code both in python and C++?

**Comment by xuezu29** [comment]
> No, it's not normal, can you show your time measuring code both in python and C++?

python
``` def time_synchronized():
    torch.cuda.synchronize() if torch.cuda.is_available() else None
    return time.time()
```
c++
`
class CTimeStamp
	{
	public:
		CTimeStamp() :
			start(0.0)
		{
			start = getTimestamp();
		}

		~CTimeStamp()
		{
			start = 0.0;
		}
		//return cost time: ms
		double getCostTimeMs()
		{
			double time = getTimestamp() - start;
			reset();
			return (time / (static_cast<double>(cv::getTickFrequency())) * 1000);
		}
		void reset()
		{
			start = getTimestamp();
		}
		void logTimeMs(const std::string& _head )
		{
			std::cout << _head + " spend time :" << getCostTimeMs() << "ms" << std::endl;
		}
	private:
		double start;
	private:
		inline double getTimestamp() {
			return static_cast<double>(cv::getTickCount());
		}
	};
`

**Comment by otakuxiang** [comment]
Well, in C++ code,have you call cudaDeviceSynchronize() function?In my case, when I second call the forward() function and use the right time measure code ,python and C++ cost the same time.

**Comment by xuezu29** [comment]
> Well, in C++ code,have you call cudaDeviceSynchronize() function?In my case, when I second call the forward() function and use the right time measure code ,python and C++ cost the same time.

 how to call cudaDeviceSynchronize() in torch C++?

**Comment by otakuxiang** [comment]
If you use Cmake ,just add cudaDeviceSynchronize() before you call time function and change 'add_executable()' to 'cuda_add_executable',otherwise you may search how to use nvcc to compile the program.

**Comment by xuezu29** [comment]
@otakuxiang 
yeah, I konw .thanks a lot!  I will try.
